icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -0,0 +1,740 @@
+        SingleUnitImpl oldUnit = null;
+        for (int i = 0, n = this.singleUnits.size(); i < n; i++) {
+            SingleUnitImpl candidate = this.singleUnits.get(i);
+            if (candidate.isCompatibleWith(singleUnit)) {
+                oldUnit = candidate;
+                break;
+            }
+        }
+
+        if (oldUnit != null) {
+            // Both dimensionalities will be positive, or both will be negative, by
+            // virtue of isCompatibleWith().
+            oldUnit.setDimensionality(oldUnit.getDimensionality() + singleUnit.getDimensionality());
+
+            return false;
+        }
+
+        // TODO: shall we just add singleUnit instead of creating a copy ??
+        this.singleUnits.add(singleUnit.clone());
+
+        // If the MeasureUnitImpl is `UMEASURE_UNIT_SINGLE` and after the appending a unit, the singleUnits are more
+        // than one singleUnit. thus means the complexity should be `UMEASURE_UNIT_COMPOUND`
+        if (this.singleUnits.size() > 1 && this.complexity == MeasureUnit.Complexity.SINGLE) {
+            this.setComplexity(MeasureUnit.Complexity.COMPOUND);
+        }
+
+        return true;
+    }
+
+    /**
+     * Transform this MeasureUnitImpl into a MeasureUnit, simplifying if possible.
+     */
+    public MeasureUnit build() {
+        return new MeasureUnit(this);
+    }
+
+    /**
+     * @return SingleUnitImpl
+     * @throws UnsupportedOperationException if the object could not be converted to SingleUnitImpl.
+     */
+    public SingleUnitImpl getSingleUnitImpl() {
+        if (this.singleUnits.size() == 0) {
+            return new SingleUnitImpl();
+        }
+        if (this.singleUnits.size() == 1) {
+            return this.singleUnits.get(0).clone();
+        }
+
+        throw new UnsupportedOperationException();
+    }
+
+    public String getIdentifier() {
+        if (this.identifier != null) {
+            return this.identifier;
+        }
+
+        this.serialize();
+        return identifier;
+    }
+
+    public MeasureUnit.Complexity getComplexity() {
+        return complexity;
+    }
+
+    public void setComplexity(MeasureUnit.Complexity complexity) {
+        this.complexity = complexity;
+    }
+
+    /**
+     * Normalizes the MeasureUnitImpl and generates the identifier string in place.
+     */
+    private void serialize() {
+        if (this.getSingleUnits().size() == 0) {
+            // Dimensionless, constructed by the default constructor: no appending
+            // to this.result, we wish it to contain the zero-length string.
+            return;
+        }
+        if (this.complexity == MeasureUnit.Complexity.COMPOUND) {
+            // Note: don't sort a MIXED unit
+            Collections.sort(this.getSingleUnits(), new SingleUnitComparator());
+        }
+
+        StringBuilder result = new StringBuilder();
+        boolean beforePre = true;
+        boolean firstTimeNegativeDimension = false;
```

Comment:
I used them determine before and after `pre` and also if the `pre` in the beginning or in the middle.

Reply:
Okay.  If you think this logic is better than the C++ logic, please change the C++.  We don't want ICU4C and ICU4J to be out of sync.

Also, please fix "pre" to "per"

Reply:
Done 

for C++ https://github.com/icu-units/icu/issues/79

Reply:
I don't see a meter-square-meter unit test.

To me personally: I get a really good and positive feeling out of seeing a unit test that failed end up passing after I implement the fix. It makes what is otherwise a "theoretical fix" feel a lot more real. (Carrot!)